src/odb/src/db/dbBlock.cpp

```diff
@@ -2089,7 +2098,65 @@ void dbBlock::setDieArea(const Rect& new_area)
   }
 }
 
+void dbBlock::setDieArea(const Polygon& new_area)
+{
+  _dbBlock* block = (_dbBlock*) this;
+  block->_die_area = new_area;
+  for (auto callback : block->_callbacks) {
+    callback->inDbBlockSetDieArea(this);
+  }
+
+  dbTech* tech = getTech();
+  dbSet<dbTechLayer> tech_layers = tech->getLayers();
+  // General flow here:
+  // We have a polygon floorplan. We can represent this floorplan
+  // as a bounding rectangle with obstructions where there is empty
+  // space in the polygon.
+  //
+  // The following code runs a subtraction on the bounding box and
+  // the polygon. This generates a list of polygons where there is
+  // supposed to be empty space. These polygons are then decomposed
+  // into rectangles. Which are then used to create obstructions and
+  // placement blockages.
+  std::list<Polygon> results;
+  Polygon bounding_rect = block->_die_area.getEnclosingRect();
+  boost::geometry::difference(bounding_rect, block->_die_area, results);
+  for (odb::Polygon& blockage_area : results) {
+    std::vector<Rect> blockages;
+    decompose_polygon(blockage_area.getPoints(), blockages);
+    for (odb::Rect& blockage_rect : blockages) {
+      dbBlockage* db_blockage = dbBlockage::create(this,
+                                                   blockage_rect.xMin(),
+                                                   blockage_rect.yMin(),
+                                                   blockage_rect.xMax(),
+                                                   blockage_rect.yMax());
+      db_blockage->setIsVirtual(true);
```

Comment:
I don't think this is the correct solution to deal with tool support. ifp, pdn, grt, drt seems like these would be the only ones that are sensitive to this and we should just invest the time to fix them without this.

Reply:
I think I disagree. I think for all the tools you listed this is essentially what they have to do anyways.

Each of their tooling support relies on deriving blockage information from odb. We could duplicate that functionality in each of them with a helper API, but this feels like this solves the problem for most tools without them having to know about it.

It doesn't fully solve the problem for ifp and ppl. Which will need to be made directly aware long term.